pkg/util/tracingservice/service.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+/*
+Package tracingservice contains a gRPC service to be used for remote inflight
+span access.
+
+It is used for pulling inflight spans from all CockroachDB nodes.
+Each node will run a trace service, which serves the inflight spans from the
+local span registry on that node. Each node will also have a trace client
+dialer, which uses the nodedialer to connect to another node's trace service,
+and access its inflight spans. The trace client dialer is backed by a remote
+trace client or a local trace client, which serve as the point of entry to this
+service. Both clients support the `TraceClient` interface, which includes the
+following functionalities:
+  - GetSpanRecordings
+*/
+package tracingservice
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracingservice/tracingservicepb"
+)
+
+// Service implements the gRPC TraceServer that exchanges inflight span
+// recordings between different nodes.
+type Service struct {
+	tracer *tracing.Tracer
+}
+
+var _ tracingservicepb.TracingServer = &Service{}
+
+// NewTracingService instantiates a tracing service server.
+func NewTracingService(tracer *tracing.Tracer) *Service {
+	return &Service{tracer: tracer}
+}
+
+// GetSpanRecordings implements the tracingpb.TraceServer interface.
+func (s *Service) GetSpanRecordings(
+	_ context.Context, request *tracingservicepb.SpanRecordingRequest,
+) (*tracingservicepb.SpanRecordingResponse, error) {
+	var resp tracingservicepb.SpanRecordingResponse
+	err := s.tracer.VisitSpans(func(span *tracing.Span) error {
```

Comment:
This is really a comment for TestTracingServiceGetSpanRecordings, but just looking at this code I think the corresponding test for it should be very simple. Setup a tracer and start two or more spans with the same trace ID, and one other with a different trace ID. It's important that these two spans are not derived from one another - we want two or more entries in the tracer's map corresponding to two forked spans from the same trace both ending up on the same node. In the test we'll then ensure that the recordings for spans with the request ID are present, and the span from the other trace isn't. 

This RPC is a purely "local" one, it doesn't internally fan out to other nodes, so our tests should be similar. The current test for this guy is doing something funky -- it's setting up a remote tracer and manually importing some remote child span into a local one. I don't think this code deal with any of that, so we're testing something more than necessary (which already has coverage for the tests written for GetSpanRecordingsFromCluster).

Reply:
Makes sense, done.